src/repr/src/scalar.rs

```diff
@@ -1228,6 +1228,65 @@ impl<'a> ScalarType {
         }
     }
 
+    /// Whether the current type is contained by the given type.
+    ///
+    /// This is a variant of `base_eq` where we allow the nullability information
+    /// of the fields within a record to be more restrictive in the current type.
+    pub fn is_covered_by(&self, other: &ScalarType) -> bool {
+        use ScalarType::*;
+        match (self, other) {
+            (
+                List {
+                    element_type: l,
+                    custom_oid: oid_l,
+                },
+                List {
+                    element_type: r,
+                    custom_oid: oid_r,
+                },
+            )
+            | (
+                Map {
+                    value_type: l,
+                    custom_oid: oid_l,
+                },
+                Map {
+                    value_type: r,
+                    custom_oid: oid_r,
+                },
+            ) => l.is_covered_by(r) && oid_l == oid_r,
+
+            (Array(a), Array(b)) => a.is_covered_by(b),
+            (
+                Record {
+                    fields: fields_a,
+                    custom_oid: oid_a,
+                    custom_name: name_a,
+                },
+                Record {
+                    fields: fields_b,
+                    custom_oid: oid_b,
+                    custom_name: name_b,
+                },
+            ) => {
+                ((fields_a.is_empty() && fields_b.is_empty())
```

Comment:
I think this line is redundant. If `((fields_a.is_empty() && fields_b.is_empty())`, then `fields_a.len() == fields_b.len()` is also true. `all()` on an empty iterator is always true. https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all